File: backend/backend/Services/ProductService.cs
using backend.DTOs;
using backend.Models;
using backend.Repositories.Implementations;
using backend.Repositories.Interfaces;

namespace backend.Services
{
    public class ProductService
    {
        private readonly IProductRepository _productRepo;

        public ProductService(IProductRepository productRepo)
        {
            _productRepo = productRepo;
        }

        //public async Task<IEnumerable<Product>> GetAllProductsAsync()
        //    => await _productRepo.GetAllAsync();

        public async Task<IEnumerable<ProductDto>> GetAllProductsAsync()
        {
            var products = await _productRepo.GetAllWithImagesAsync();

            //return products.Select(p => new ProductDto
            //{
            //    Id = p.Id,
            //    Name = p.Name,
            //    Description = p.Description,
            //    Price = p.Price,
            //    DiscountPrice = p.DiscountPrice,
            //    StockQuantity = p.StockQuantity,
            //    CategoryId = p.CategoryId,

            //    ImageUrls = p.Images.Select(i => i.ImageUrl).ToList()
            //});
            return products.Select(p => new ProductDto
            {
                Id = p.Id,
                Name = p.Name,
                Description = p.Description,
                Price = p.Price,
                DiscountPrice = p.DiscountPrice,
                StockQuantity = p.StockQuantity,
                CategoryId = p.CategoryId,
                AverageRating = p.AverageRating,
                CategoryName = p.Category?.Name,
                Images = p.Images.Select(i => new ProductImageDto
                {
                    Id = i.Id,
                    ImageUrl = i.ImageUrl,
                    IsPrimary = i.IsPrimary,
                    DisplayOrder = i.DisplayOrder
                }).ToList()
            });
        }
        //public async Task<Product?> GetProductByIdAsync(int id)
        //    => await _productRepo.GetByIdAsync(id);


        public async Task<ProductDto?> GetProductByIdAsync(int id)
        {
            var product = await _productRepo.GetByIdAsync(id);

            if (product == null) return null;

            //return new ProductDto
            //{
            //    Id = product.Id,
            //    Name = product.Name,
            //    Description = product.Description,
            //    Price = product.Price,
            //    DiscountPrice = product.DiscountPrice,
            //    StockQuantity = product.StockQuantity,
            //    CategoryId = product.CategoryId,

            //    ImageUrls = product.Images
            //        .Select(i => i.ImageUrl)
            //        .ToList()
            //};
            return new ProductDto
            {
                Id = product.Id,
                Name = product.Name,
                Description = product.Description,
                Price = product.Price,
                DiscountPrice = product.DiscountPrice,
                StockQuantity = product.StockQuantity,
                CategoryId = product.CategoryId,
                AverageRating = product.AverageRating,
                CategoryName = product.Category?.Name,
                Images = product.Images.Select(i => new ProductImageDto
                {
                    Id = i.Id,
                    ImageUrl = i.ImageUrl,
                    IsPrimary = i.IsPrimary,
                    DisplayOrder = i.DisplayOrder
                }).ToList()
            };
        }

        public async Task<Product> CreateProductAsync(Product product)
            => await _productRepo.AddAsync(product);
    }
}

File: frontend/src/app/core/services/product.service.ts
import { Injectable, inject } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable, map } from "rxjs";
import {
  Product,
  ProductFilter,
  PaginatedResult,
  Category,
  Review,
  CreateProductDto,
  CreateReviewDto,
} from "../models/product.model";
import { environment } from "../../../environments/environment";

@Injectable({ providedIn: "root" })
export class ProductService {
  private readonly API = `${environment.apiUrl}/api`;
  private http = inject(HttpClient);

  // GET /api/Products — backend returns array, we wrap into PaginatedResult
  getProducts(filter?: ProductFilter): Observable<PaginatedResult<Product>> {
    return this.http.get<any[]>(`${this.API}/Products`).pipe(
      map((data) => {
        const arr = Array.isArray(data) ? data : [];
        const products: Product[] = arr.map((p) => this.mapProduct(p));

        // Client-side filter by category
        const filtered = filter?.categoryId
          ? products.filter((p) => p.categoryId === filter.categoryId)
          : products;

        const page = filter?.page ?? 1;
        const limit = filter?.limit ?? 12;
        const totalItems = filtered.length;
        const totalPages = Math.ceil(totalItems / limit) || 1;
        const start = (page - 1) * limit;
        const items = filtered.slice(start, start + limit);

        return { items, totalItems, totalPages, currentPage: page };
      }),
    );
  }

  // GET /api/Products/{id}
  getProduct(id: number): Observable<Product> {
    return this.http
      .get<any>(`${this.API}/Products/${id}`)
      .pipe(map((p) => this.mapProduct(p)));
  }

  getByCategory(categoryId: number): Observable<Product[]> {
    return this.http
      .get<any[]>(`${this.API}/Products`)
      .pipe(
        map((arr) =>
          arr
            .filter((p) => p.categoryId === categoryId)
            .map((p) => this.mapProduct(p)),
        ),
      );
  }

  getFeatured(): Observable<Product[]> {
    return this.http
      .get<any[]>(`${this.API}/Products`)
      .pipe(map((arr) => arr.slice(0, 8).map((p) => this.mapProduct(p))));
  }

  search(
    q: string,
    filter?: ProductFilter,
  ): Observable<PaginatedResult<Product>> {
    return this.http.get<any[]>(`${this.API}/Products`).pipe(
      map((data) => {
        const arr = Array.isArray(data) ? data : [];
        const all = arr.map((p) => this.mapProduct(p));
        const filtered = all.filter(
          (p) =>
            p.name.toLowerCase().includes(q.toLowerCase()) ||
            p.description?.toLowerCase().includes(q.toLowerCase()),
        );
        const page = filter?.page ?? 1;
        const limit = filter?.limit ?? 12;
        const totalItems = filtered.length;
        const totalPages = Math.ceil(totalItems / limit) || 1;
        const start = (page - 1) * limit;
        return {
          items: filtered.slice(start, start + limit),
          totalItems,
          totalPages,
          currentPage: page,
        };
      }),
    );
  }

  // Map backend DTO → frontend Product model
  private mapProduct(p: any): Product {
    return {
      id: p.id,
      name: p.name,
      description: p.description ?? "",
      price: p.price,
      discountPrice: p.discountPrice,
      stockQuantity: p.stockQuantity ?? 0,
      sku: p.sku ?? p.sKU ?? "",
      categoryId: p.categoryId,
      sellerId: p.sellerId,
      createdAt: p.createdAt,
      updatedAt: p.updatedAt,
      isActive: p.isActive ?? true,
      viewCount: p.viewCount ?? 0,
      averageRating: p.averageRating ?? 0,
      category: p.categoryName
        ? {
            id: p.categoryId,
            name: p.categoryName,
            description: "",
            isActive: true,
          }
        : p.category,
      // Handle both: images[] of objects  OR  imageUrls[] of strings
      images: p.images?.length
        ? p.images
        : (p.imageUrls ?? []).map((url: string, i: number) => ({
            id: i,
            productId: p.id,
            imageUrl: url,
            isPrimary: i === 0,
            displayOrder: i,
          })),
    };
  }

  // POST /api/Products
  createProduct(data: FormData): Observable<Product> {
    return this.http.post<Product>(`${this.API}/Products`, data);
  }

  // PUT /api/Products/{id}
  updateProduct(id: number, data: any): Observable<Product> {
    return this.http.put<Product>(`${this.API}/Products/${id}`, data);
  }

  // DELETE /api/Products/{id}
  deleteProduct(id: number): Observable<any> {
    return this.http.delete(`${this.API}/Products/${id}`);
  }

  uploadImages(productId: number, images: FormData): Observable<any> {
    return this.http.post(`${this.API}/Products/${productId}/images`, images);
  }

  deleteImage(imageId: number): Observable<any> {
    return this.http.delete(`${this.API}/Products/images/${imageId}`);
  }

  // GET /api/Reviews/product/{productId}
  getReviews(productId: number): Observable<Review[]> {
    return this.http.get<Review[]>(`${this.API}/Reviews/product/${productId}`);
  }

  // POST /api/Reviews
  addReview(review: CreateReviewDto): Observable<Review> {
    return this.http.post<Review>(`${this.API}/Reviews`, review);
  }

  // DELETE /api/Reviews/{id}
  deleteReview(id: number): Observable<any> {
    return this.http.delete(`${this.API}/Reviews/${id}`);
  }

  // GET /api/Categories
  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.API}/Categories`);
  }

  getCategory(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.API}/Categories/${id}`);
  }

  createCategory(data: Partial<Category>): Observable<Category> {
    return this.http.post<Category>(`${this.API}/Categories`, data);
  }

  updateCategory(id: number, data: Partial<Category>): Observable<Category> {
    return this.http.put<Category>(`${this.API}/Categories/${id}`, data);
  }

  deleteCategory(id: number): Observable<any> {
    return this.http.delete(`${this.API}/Categories/${id}`);
  }
}
